README
NAME
    WebService::PayPal::PaymentsAdvanced - A simple wrapper around the
    PayPal Payments Advanced web service

VERSION
    version 0.000007

SYNOPSIS
        use WebService::PayPal::PaymentsAdvanced;
        my $payments = WebService::PayPal::PaymentsAdvanced->new(
            {
                password => 'seekrit',
                user     => 'username',
                vendor   => 'somevendor',
            }
        );

        my $response = $payments->create_secure_token(
            {
                AMT            => 100,
                TRXTYPE        => 'S',
                BILLINGTYPE    => 'MerchantInitiatedBilling',
                CANCELURL      => 'https://example.com/cancel',
                ERRORURL       => 'https://example.com/error',
                L_BILLINGTYPE0 => 'MerchantInitiatedBilling',
                NAME           => 'Chuck Norris',
                RETURNURL      => 'https://example.com/return',
            }
        );

        my $uri = $response->hosted_form_uri;

        # Store token data for later use.  You'll need to implement this yourself.
        $foo->freeze_token_data(
            token    => $response->secure_token,
            token_id => $response->secure_token_id,
        );

        # Later, when PayPal returns a silent POST or redirects the user to your
        # return URL:

        my $redirect_response = $payments->get_response_from_redirect(
            ip_address => $ip,
            params     => $params,
        );

        # Fetch the tokens from the original request. You'll need to implement
        # this yourself.

        my $thawed = $foo->get_thawed_tokens(...);

        # Don't do anything until you're sure the tokens are ok.
        if (   $thawed->secure_token ne $redirect->secure_token
            || $thawed->secure_token_id ne $response->secure_token_id ) {
            die 'Fraud!';
        }

        # Everything looks good.  Carry on!

    print $response->secure_token;

DESCRIPTION
    BETA BETA BETA. The interface is still subject to change.

    This is a wrapper around the "PayPal Payments Advanced" (AKA "PayPal
    Payflow Link") hosted forms. This code does things like facilitating
    secure token creation, providing an URL which you can use to insert an
    hosted_form into your pages and processing the various kinds of response
    you can get from PayPal.

    We also use various exception classes to make it easier for you to
    decide how to handle the parts that go wrong.

OBJECT INSTANTIATION
    The following parameters can be supplied to "new()" when creating a new
    object.

  Required Parameters
   password
    The value of the "password" field you use when logging in to the Payflow
    Manager. (You'll probably want to create a specific user just for API
    calls).

   user
    The value of the "user" field you use when logging in to the Payflow
    Manager.

   vendor
    The value of the "vendor" field you use when logging in to the Payflow
    Manager.

  Optional Parameters
   partner
    The value of the "partner" field you use when logging in to the Payflow
    Manager. Defaults to "PayPal".

   payflow_pro_uri
    The hostname for the Payflow Pro API. This is where token creation
    requests get directed. This already has a sensible (and correct)
    default, but it is settable so that you can more easily mock API calls
    when testing.

   payflow_link_uri
    The hostname for the Payflow Link website. This is the hosted service
    where users will enter their payment information. This already has a
    sensible (and correct) default, but it is settable in case you want to
    mock it while testing.

   production_mode
    This is a "Boolean". Set this to "true" if when you are ready to process
    real transactions. Defaults to "false".

   ua
    You may provide your own UserAgent, but it must be of the LWP::UserAgent
    family. If you do provide a UserAgent, be sure to set a sensible timeout
    value.

    This can be useful for debugging. You'll be able to get detailed
    information about the network calls which are being made.

        use LWP::ConsoleLogger::Easy qw( debug_ua );
        use LWP::UserAgent;
        use WebService::PayPal::PaymentsAdvanced;

        my $ua = LWP::UserAgent;
        debug_ua($ua);

        my $payments
            = WebService::PayPal::PaymentsAdvanced->new( ua => $ua, ... );

        # Now fire up a console and watch your network activity.

    Check the tests which accompany this distribution for an example of how
    to mock API calls using Test::LWP::UserAgent.

   validate_hosted_form_uri
    "Boolean". If enabled, this module will attempt to GET the uri which
    you'll be providing to the end user. This can help you identify issues
    on the PayPal side. This is helpful because you'll be able to log
    exceptions thrown by this method and deal with them accordingly. If you
    disable this option, you'll need to rely on end users to report issues
    which may exist within PayPal's hosted pages. Defaults to "true".

   verbose
    "Boolean". Sets "VERBOSITY=HIGH" on all transactions if enabled.
    Defaults to "true".

  Methods
   create_secure_token
    Create a secure token which you can use to create a hosted form uri.
    Returns a WebService::PayPal::PaymentsAdvanced::Response::SecureToken
    object.

        use WebService::PayPal::PaymentsAdvanced;
        my $payments = WebService::PayPal::PaymentsAdvanced->new(...);

        my $response = $payments->create_secure_token(
            {
                AMT            => 100,
                TRXTYPE        => 'S',
                BILLINGTYPE    => 'MerchantInitiatedBilling',
                CANCELURL      => 'https://example.com/cancel',
                ERRORURL       => 'https://example.com/error',
                L_BILLINGTYPE0 => 'MerchantInitiatedBilling',
                NAME           => 'Chuck Norris',
                RETURNURL      => 'https://example.com/return'
            }
        );

        print $response->secure_token;

   get_response_from_redirect
    This method can be used to parse responses from PayPal to your return
    URL. It's essentially a wrapper around
    WebService::PayPal::PaymentsAdvanced::Response::FromRedirect. Returns a
    WebService::PayPal::PaymentsAdvanced::Response object.

        my $response = $payments->get_response_from_redirect(
            params     => $params,
        );
        print $response->message;

   get_response_from_silent_post
    This method can be used to validate responses from PayPal to your silent
    POST url. If you provide an ip_address parameter, it will be validated
    against a list of known IPs which PayPal provides. You're encouraged to
    provide an IP address in order to prevent spoofing of payment responses.
    See WebService::PayPal::PaymentsAdvanced::Response::FromSilentPOST for
    more information on this behaviour.

    This method returns a
    WebService::PayPal::PaymentsAdvanced::Response::FromSilentPost::PayPal
    object for PayPal transactions. It returns a
    WebService::PayPal::PaymentsAdvanced::Response::FromSilentPost::CreditCa
    rd object for credit card transactions. You can either inspect the class
    return to you or use the "is_credit_card_transaction" or
    "is_paypal_transaction" methods to learn which method the customer paid
    with. Both methods return a "Boolean".

        my $response = $payments->get_response_from_redirect(
            ip_address => $ip,
            params     => $params,
        );
        print $response->message. "\n";
        if ( $response->is_credit_card_transaction ) {
            print $response->card_type, q{ }, $response->card_expiration;
        }

   post
    Generic method to post arbitrary params to PayPal. Requires a "HashRef"
    of parameters and returns a
    WebService::PayPal::PaymentsAdvanced::Response object. Any lower case
    keys will be converted to upper case before this response is sent.

        use WebService::PayPal::PaymentsAdvanced;
        my $payments = WebService::PayPal::PaymentsAdvanced->new(...);

        my $response = $payments->post( { TRXTYPE => 'V', ORIGID => $pnref, } );
        say $response->message;

        # OR
        my $response = $payments->post( { trxtype => 'V', origid => $pnref, } );

   capture_delayed_transaction( $ORIGID )
    Captures a sale which you have previously authorized. Requires the ID of
    the original transaction. Returns a response object.

   auth_from_credit_card_reference_transaction( $ORIGID, $amount )
    Process a authorization based on a reference transaction from a credit
    card. Requires 2 arguments: an ORIGID from a previous credit card
    transaction and an amount.

        use WebService::PayPal::PaymentsAdvanced;
        my $payments = WebService::PayPal::PaymentsAdvanced->new(...);

        my $response = $payments->auth_from_credit_card_reference_transaction(
            'BFOOBAR', 1.50'
        );
        say $response->message;

   sale_from_credit_card_reference_transaction( $ORIGID, $amount )
    Process a sale based on a reference transaction from a credit card. See
    Requires 2 arguments: an ORIGID from a previous credit card transaction
    and an amount.

        use WebService::PayPal::PaymentsAdvanced;
        my $payments = WebService::PayPal::PaymentsAdvanced->new(...);

        my $response = $payments->sale_from_credit_card_reference_transaction(
            'BFOOBAR', 1.50'
        );
        say $response->message;

   auth_from_paypal_reference_transaction( $BAID, $amount, $currency )
    Process an authorization based on a reference transaction from PayPal.
    Requires 3 arguments: a BAID from a previous PayPal transaction, an
    amount and a currency.

        use WebService::PayPal::PaymentsAdvanced;
        my $payments = WebService::PayPal::PaymentsAdvanced->new(...);

        my $response = $payments->auth_from_paypal_reference_transaction(
            'B-FOOBAR', 1.50, 'USD'
        );
        say $response->message;

   sale_from_paypal_reference_transaction( $BAID, $amount, $currency )
    Process a sale based on a reference transaction from PayPal. Requires 3
    arguments: a BAID from a previous PayPal transaction, an amount and a
    currency.

        use WebService::PayPal::PaymentsAdvanced;
        my $payments = WebService::PayPal::PaymentsAdvanced->new(...);

        my $response = $payments->sale_from_paypal_reference_transaction(
            'B-FOOBAR', 1.50, 'USD'
        );
        say $response->message;

   inquiry_transaction( $ORIGID )
    Performs a transaction inquiry on a previously submitted transaction.
    Requires the ID of the original transaction. Returns a response object.

   void_transaction( $ORIGID )
    Voids a previous transaction. Requires the ID of the transaction to
    void. Returns a response object.

AUTHOR
    Olaf Alders <olaf@wundercounter.com>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2015 by MaxMind, Inc..

    This is free software, licensed under:

      The Artistic License 2.0 (GPL Compatible)



Changes
Revision history for WebService-PayPal-PaymentsAdvanced

{{$NEXT}}

0.000028 2022-01-01

  - Update 12/21 expiration dates in mocked data to 12/31.

0.000027 2021-06-28

  - The transaction_time method on responses will now return undef if
    the TRANSTIME parameter is missing or undefined. Previously an
    exception would be thrown.

0.000026 2020-04-29

   - You may now use a user or password object that stringifies to the
     actual user and password.

0.000025 2018-02-05

   - You may now set the MODE parameter on the hosted_form_uri by
     setting hosted_form_mode to "LIVE" or "TEST".

0.000024 2017-12-21

    - Add new PayPal IPs to silent POST verification.

0.000023 2017-12-15
    - Accept 0000 as the last four digits of a credit card.

0.000022 2017-05-11
    - Provide an option to retry hosted form validation HTTP requests on 5xx
      errors. Also provide a callback function we call when encountering such
      errors prior to retrying.

0.000021 2016-06-13
    - Use correct throw method for HTTP error.

0.000020 2016-06-13
    - Update notify.paypal.com IP addresses.

0.000019 2016-06-09
    - Always add the response content to the HTTP error exception messages as
      many fake LWP::UserAgent 500s do not set any distinguishing headers.

0.000018 2016-05-18
    - The connection timeout was increased to 60 seconds as requests to the
      service frequently run 20-30 seconds.
    - Previously, this API would throw
      WebService::PayPal::PaymentsAdvanced::Error::HTTP exceptions with
      messages indicating that there was an HTTP error when the request failed
      for other reasons such as a timeout. The message should now more
      accurately reflect the cause of the exception.

0.000017 2016-02-16
    - Successful refunds now return a
      WebService::PayPal::PaymentsAdvanced::Response::Credit response object.

0.000016 2016-02-09
    - Successful voids now return a
      WebService::PayPal::PaymentsAdvanced::Response::Void response object.

0.000015 2016-01-22
    - Make less assumptions about inquiry_transaction().  This is a breaking
      change.

0.000014 2015-12-08
    - *_reference_transaction() methods now accept an additional, optional
      argument for parameters which are to be passed through directly to
      PayPal.

0.000013 2015-12-03
    - PaymentsAdvanced constructor now takes an optional arrayref of result
      codes to treat as non-fatal.

0.000012 2015-11-23
    - capture_delayed_transaction() now optionally accepts an amount.

0.000010 2015-11-11
    - Response objects no longer assume that PPREF will always be present.

0.000009 2015-11-04
    - WebService::PayPal::PaymentsAdvanced::Error::HTTP now provides
      request_uri. WebService::PayPal::PaymentsAdvanced::Error::HostedForm
      provides request_uri and http_status.

0.000008  2015-10-27 17:58:19-04:00 America/Toronto
    - Fixes an error when adding query params to a URI object

0.000007  2015-10-16 21:05:02-04:00 America/Toronto
    - Adds auth_from_credit_card_reference_transaction()
    - Adds auth_from_paypal_reference_transaction()

0.000006  2015-09-14 10:20:56-04:00 America/Toronto
    - Adds built-in mocking for testing and running your own local mocked
      PayPal services
    - Moves away from using one generic response object to using various, more
      specific response objects

0.000005  2015-08-07 10:52:50-04:00 America/Toronto
    - Adds post() and void_transaction methods to PaymentsAdvanced.pm
    - Makes use of HashRefs more consistent in API
    - Adds new methods: capture_delayed_transaction(), transaction_status() and
      sale_from_reference_transaction
    - Adds "verbose" attribute to constructor

0.000004  2015-07-24 17:34:27-04:00 America/Toronto
    - Bumps up minimum version of Moo to 1.004005

0.000003  2015-07-22 13:04:12-04:00 America/Toronto
    - Adds LWP::Protocol::https to deps
    - Removes IP address validation on redirect

0.000002  2015-07-20 14:41:56-04:00 America/Toronto
    - Don't support Perls earlier than 5.14

0.000001  2015-07-17 18:27:22-04:00 America/Toronto
    - Initial release.


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.022.
use strict;
use warnings;

use 5.014;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "A simple wrapper around the PayPal Payments Advanced web service",
  "AUTHOR" => "Olaf Alders <olaf\@wundercounter.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "WebService-PayPal-PaymentsAdvanced",
  "EXE_FILES" => [
    "bin/mock-payflow-link",
    "bin/mock-payflow-pro"
  ],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.014",
  "NAME" => "WebService::PayPal::PaymentsAdvanced",
  "PREREQ_PM" => {
    "Const::Fast" => 0,
    "Data::GUID" => 0,
    "DateTime" => 0,
    "DateTime::Format::MySQL" => 0,
    "DateTime::TimeZone" => 0,
    "HTTP::Message::PSGI" => 0,
    "HTTP::Status" => 0,
    "LWP::Protocol::https" => 0,
    "LWP::UserAgent" => 0,
    "List::AllUtils" => 0,
    "Mojolicious::Lite" => 0,
    "Moo" => "1.004005",
    "Moo::Role" => 0,
    "MooX::StrictConstructor" => 0,
    "Net::Works::Address" => 0,
    "Net::Works::Network" => 0,
    "Test::LWP::UserAgent" => 0,
    "Throwable::Error" => 0,
    "Type::Params" => 0,
    "Type::Utils" => 0,
    "Types::Common::Numeric" => 0,
    "Types::Common::String" => 0,
    "Types::Standard" => 0,
    "Types::URI" => 0,
    "URI" => "1.69",
    "URI::FromHash" => 0,
    "URI::QueryParam" => 0,
    "Web::Scraper" => 0,
    "feature" => 0,
    "namespace::autoclean" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "HTTP::Response" => 0,
    "LWP::ConsoleLogger::Easy" => 0,
    "Mozilla::CA" => 20130114,
    "Path::Tiny" => 0,
    "Scalar::Util" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "0.96",
    "Test::RequiresInternet" => 0,
    "Try::Tiny" => 0,
    "lib" => 0,
    "overload" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.000028",
  "test" => {
    "TESTS" => "t/*.t t/PaymentsAdvanced/*.t t/PaymentsAdvanced/Mocker/*.t t/PaymentsAdvanced/Response/*.t t/PaymentsAdvanced/Response/FromSilentPOST/*.t"
  }
);


my %FallbackPrereqs = (
  "Const::Fast" => 0,
  "Data::GUID" => 0,
  "DateTime" => 0,
  "DateTime::Format::MySQL" => 0,
  "DateTime::TimeZone" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "HTTP::Message::PSGI" => 0,
  "HTTP::Response" => 0,
  "HTTP::Status" => 0,
  "LWP::ConsoleLogger::Easy" => 0,
  "LWP::Protocol::https" => 0,
  "LWP::UserAgent" => 0,
  "List::AllUtils" => 0,
  "Mojolicious::Lite" => 0,
  "Moo" => "1.004005",
  "Moo::Role" => 0,
  "MooX::StrictConstructor" => 0,
  "Mozilla::CA" => 20130114,
  "Net::Works::Address" => 0,
  "Net::Works::Network" => 0,
  "Path::Tiny" => 0,
  "Scalar::Util" => 0,
  "Test::Fatal" => 0,
  "Test::LWP::UserAgent" => 0,
  "Test::More" => "0.96",
  "Test::RequiresInternet" => 0,
  "Throwable::Error" => 0,
  "Try::Tiny" => 0,
  "Type::Params" => 0,
  "Type::Utils" => 0,
  "Types::Common::Numeric" => 0,
  "Types::Common::String" => 0,
  "Types::Standard" => 0,
  "Types::URI" => 0,
  "URI" => "1.69",
  "URI::FromHash" => 0,
  "URI::QueryParam" => 0,
  "Web::Scraper" => 0,
  "feature" => 0,
  "lib" => 0,
  "namespace::autoclean" => 0,
  "overload" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
